api/v1alpha1/backend_types.go

```diff
@@ -0,0 +1,157 @@
+//
+// +kubebuilder:validation:XValidation:rule="(has(self.socketAddress) || has(self.unixDomainSocketAddress))",message="one of socketAddress or unixDomainSocketAddress must be specified"
+// +kubebuilder:validation:XValidation:rule="(has(self.socketAddress) && !has(self.unixDomainSocketAddress)) || (!has(self.socketAddress) && has(self.unixDomainSocketAddress))",message="only one of socketAddress or unixDomainSocketAddress can be specified"
+// +kubebuilder:validation:XValidation:rule="((has(self.socketAddress) && (self.type == 'FQDN' || self.type == 'IPv4')) || has(self.unixDomainSocketAddress) && self.type == 'UDS')",message="if type is FQDN or IPv4, socketAddress must be set; if type is UDS, unixDomainSocketAddress must be set"
+// +notImplementedHide
+type BackendAddress struct {
+	// Type is the the type name of the backend address: FQDN, UDS, IPv4
+	Type AddressType `json:"type"`
+
+	// SocketAddress defines a FQDN or IPv4 address
+	//
+	// +optional
+	SocketAddress *SocketAddress `json:"socketAddress,omitempty"`
+
+	// UnixDomainSocketAddress defines the unix domain socket path
+	//
+	// +optional
+	UnixDomainSocketAddress *UnixDomainSocketAddress `json:"unixDomainSocketAddress,omitempty"`
+}
+
+// SocketAddress describes TCP/UDP socket address, corresponding to Envoy's SocketAddress
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-socketaddress
+// +notImplementedHide
+type SocketAddress struct {
+	// Address defines to the FQDN or IP address of the backend service.
+	Address string `json:"address"`
+
+	// Port defines to the port of of the backend service.
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port int32 `json:"port"`
+
+	// Protocol defines to the the transport protocol to use for communication with the backend.
+	//
+	// +optional
+	Protocol *ProtocolType `json:"protocol,omitempty"`
+}
+
+// UnixDomainSocketAddress describes TCP/UDP unix domain socket address, corresponding to Envoy's Pipe
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-pipe
+// +notImplementedHide
+type UnixDomainSocketAddress struct {
+	Path string `json:"path"`
+}
+
+// BackendSpec describes the desired state of BackendSpec.
+// +notImplementedHide
+type BackendSpec struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=4
+	// +kubebuilder:validation:XValidation:rule="self.all(f, f.type == 'FQDN') || !self.exists(f, f.type == 'FQDN')",message="FQDN addresses cannot be mixed with other address types"
+	BackendAddresses []BackendAddress `json:"addresses,omitempty"`
+
+	// ApplicationProtocol defines the application protocol to be used, e.g. HTTP2.
+	//
+	// +optional
+	ApplicationProtocol *ApplicationProtocolType `json:"applicationProtocol,omitempty"`
```

Comment:
can we do w/o this field in the init PR and rely on info from the xRoute for the protocol ?

Reply:
I think that it's not currently possible to infer if HTTP1 or HTTP2 should be used upstream in HTTP routes. So, we need something similar to K8s service port application protocols.

Reply:
suggest making it a list if we want to add it in the first iteration
relates to https://github.com/kubernetes/enhancements/issues/3726

reg appProtocol naming, can we reuse anything from upstream K8s https://github.com/kubernetes/enhancements/pull/3727 ?

Reply:
> suggest making it a list

Can you explain why we need a list based on the KEP?

Reply:
+1 to naming suggestion

Reply:
also prefer `appProtocols` over `applicationProocols`

Reply:
ok, makes sense. Pluralized it.